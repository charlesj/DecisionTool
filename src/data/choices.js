import { fromJS } from 'immutable'
import shortid from 'shortid'
import { paths } from './paths'

export const addChoice = (state) => {
  const choiceId = shortid()
  return state.setIn(paths.choice(choiceId), fromJS({ name: `choice ${choiceId}` }))
}

export const updateChoiceName = (state, choiceId, name) => {
  return state.setIn([...paths.choice(choiceId), 'name'], name)
}

export const updateFactorScore = (state, choiceId, factorId, score) => {
  return state.setIn(paths.choiceFactorScore(choiceId, factorId), score)
}

export const removeChoice = (state, choiceId) => {
  return state.deleteIn(paths.choice(choiceId))
}
