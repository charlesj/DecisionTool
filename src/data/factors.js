import { fromJS } from 'immutable'
import shortid from 'shortid'
import { paths } from './paths'

export const addFactor = (state) => {
  const factorId = shortid()
  return state.setIn(paths.factor(factorId), fromJS({ factorId, name: `factor ${factorId}` }))
}

export const updateFactorName = (state, factorId, name) => {
  return state.setIn([...paths.factor(factorId), 'name'], name)
}

export const updateFactorWeight = (state, factorId, weight) => {
  return state.setIn([...paths.factor(factorId), 'weight'], weight)
}

export const removeFactor = (state, factorId) => {
  return state.deleteIn(paths.factor(factorId))
}
