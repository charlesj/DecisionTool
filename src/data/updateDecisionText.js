import { paths } from './paths'

export const updateDecisionText = (currState, decisionText) => {
  return currState.setIn(paths.decisionText, decisionText)
}
