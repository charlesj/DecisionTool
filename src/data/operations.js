import { updateDecisionText } from './updateDecisionText'
import * as factors from './factors'
import * as choices from './choices'

export const getOperations = (currentState, updateState) => {
  return {
    updateDecisionText: (decisionText) => updateState(updateDecisionText(currentState, decisionText)),
    addFactor: () => updateState(factors.addFactor(currentState)),
    updateFactorName: (factorId, name) => updateState(factors.updateFactorName(currentState, factorId, name)),
    updateFactorWeight: (factorId, weight) => updateState(factors.updateFactorWeight(currentState, factorId, weight)),
    removeFactor: (factorId) => updateState(factors.removeFactor(currentState, factorId)),
    addChoice: () => updateState(choices.addChoice(currentState)),
    updateChoiceName: (choiceId, name) => updateState(choices.updateChoiceName(currentState, choiceId, name)),
    updateChoiceFactorScore: (choiceId, factorId, score) => updateState(choices.updateFactorScore(currentState, choiceId, factorId, score)),
    removeChoice: (choiceId) => updateState(choices.removeChoice(currentState, choiceId))
  }
}
