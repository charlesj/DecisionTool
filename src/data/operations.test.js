import { fromJS } from 'immutable'
import { updateDecisionText } from './updateDecisionText'
import * as factors from './factors'
import * as choices from './choices'
import { getOperations } from './operations'

jest.mock('./factors')
jest.mock('./choices')
jest.mock('./updateDecisionText')

const updateState = jest.fn()
const initState = fromJS({})

const mop = (op, opName) => op.mockImplementation((state, ...rest) => state.setIn([opName], fromJS(rest)))

let ops
beforeEach(() => {
  mop(updateDecisionText, 'updateDecisionText')
  mop(factors.addFactor, 'addFactor')
  mop(factors.removeFactor, 'removefactor')
  mop(factors.updateFactorName, 'update factor name')
  mop(factors.updateFactorWeight, 'factor weight')
  mop(choices.addChoice, 'add choice')
  mop(choices.updateChoiceName, 'update choice name')
  mop(choices.updateFactorScore, 'choice factor score')
  mop(choices.removeChoice, 'remove choice')
  ops = getOperations(initState, updateState)
})

test('getOperations includes updateDecisionText', () => {
  ops.updateDecisionText('new decision text')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('add factor', () => {
  ops.addFactor()
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('remove factor', () => {
  ops.removeFactor('factor id')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('update factor name', () => {
  ops.updateFactorName('factorId', 'factor name')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('update factor weight', () => {
  ops.updateFactorWeight('factorId', 'factor weight')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('add choice', () => {
  ops.addChoice()
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('update choice name', () => {
  ops.updateChoiceName('choice-id', 'new name')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('update choice factor score', () => {
  ops.updateChoiceFactorScore('choice-id', 'factor-id', 'score')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})

test('remove choice', () => {
  ops.removeChoice('choice-id')
  expect(updateState.mock.calls[0][0]).toMatchSnapshot()
})
