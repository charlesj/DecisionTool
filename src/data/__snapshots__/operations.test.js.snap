// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add choice 1`] = `
Immutable.Map {
  "add choice": Immutable.List [],
}
`;

exports[`add factor 1`] = `
Immutable.Map {
  "addFactor": Immutable.List [],
}
`;

exports[`getOperations includes updateDecisionText 1`] = `
Immutable.Map {
  "updateDecisionText": Immutable.List [
    "new decision text",
  ],
}
`;

exports[`remove choice 1`] = `
Immutable.Map {
  "remove choice": Immutable.List [
    "choice-id",
  ],
}
`;

exports[`remove factor 1`] = `
Immutable.Map {
  "removefactor": Immutable.List [
    "factor id",
  ],
}
`;

exports[`update choice factor score 1`] = `
Immutable.Map {
  "choice factor score": Immutable.List [
    "choice-id",
    "factor-id",
    "score",
  ],
}
`;

exports[`update choice name 1`] = `
Immutable.Map {
  "update choice name": Immutable.List [
    "choice-id",
    "new name",
  ],
}
`;

exports[`update factor name 1`] = `
Immutable.Map {
  "update factor name": Immutable.List [
    "factorId",
    "factor name",
  ],
}
`;

exports[`update factor weight 1`] = `
Immutable.Map {
  "factor weight": Immutable.List [
    "factorId",
    "factor weight",
  ],
}
`;
